ext/pdo/pdo_dbh.c

```diff
@@ -678,11 +678,20 @@ PHP_METHOD(PDO, inTransaction)
 }
 /* }}} */
 
-/* TODO: Make distinction between numeric and non-numeric strings */
-#define PDO_LONG_PARAM_CHECK \
-	if (Z_TYPE_P(value) != IS_LONG && Z_TYPE_P(value) != IS_STRING && Z_TYPE_P(value) != IS_FALSE && Z_TYPE_P(value) != IS_TRUE) { \
-		zend_type_error("Attribute value must be of type int for selected attribute, %s given", zend_zval_type_name(value)); \
-		return false; \
+#define PDO_GET_LONG_PARAM(zval) \
+	switch (Z_TYPE_P(zval)) { \
+		case IS_LONG: \
+		case IS_TRUE: \
+		case IS_FALSE: \
+			lval = zval_get_long(zval); \
+			break; \
+		case IS_STRING: \
+			if (IS_LONG == is_numeric_str_function(Z_STR_P(zval), &lval, NULL)) { \
+				break; \
+			} \
+			/* fallthrough */ \
+		default: \
+			zend_type_error("Attribute value must be of type int for selected attribute, %s given", zend_zval_type_name(zval)); \
```

Comment:
Isn't this missing a return? I would also prefer this to be converted into a function

```
if (!pdo_get_long_param(&lval, value)) {
    return false;
}
```